packages/stable/src/cogniteClient.ts

```diff
@@ -172,14 +174,33 @@ export default class CogniteClient extends BaseCogniteClient {
   private geospatialApi?: GeospatialAPI;
   private documentsApi?: DocumentsAPI;
 
+  constructor(options: ClientOptions, apiVersion: CogniteAPIVersion = 'v1') {
+    super(options, apiVersion);
+    this.checkCoreVersionMatch();
+  }
+
   protected get version() {
     return version;
   }
 
+  protected get coreVersion() {
+    // @ts-ignore linter is complaining even though version is protected and can be accessed
+    return super.version;
+  }
+
   protected getRetryValidator(): RetryValidator {
     return retryValidator;
   }
 
+  private checkCoreVersionMatch() {
+    // check if sdk-core version matches with the one in the dependencies
+    if (!dependencies['@cognite/sdk-core'].includes(this.coreVersion)) {
+      console.warn(
+        `VERSION MISMATCH! \n The resolved @cognite/sdk-core version ${this.coreVersion} doesn't match the required version ${dependencies['@cognite/sdk-core']} from @cognite/sdk. This might lead to unexpected behavior and bugs.`
+      );
+    }
+  }
+
```

Comment:
I'm not sure if this will work as you expect:

If a user installs `@cognite/sdk` and also an incompatible version of `@cognite/sdk-core` the core package will exist multiple places in the dependency tree due to how version resolution is done. With NPM you will not easily be able to put an incompatible version for this import, while with yarn you can force it with resolutions.

For the (normal?) situations where the user accidentally installs an incompatible version but don't override the transitive version, this logic will not catch that situation.

Also note that you probably want to do a semver check here. The value in package.json is a range, while the `coreVersion` is a specific version. While the `includes` work at the point of release, it will break on next patch release even though it would be considered a compatible release (and version resolution will pick it).

Have you seen any other projects that tries to solve the same problem? Maybe there is some useful inspiration around.

Reply:
@henrist The monorepos had special tsconfig setup and it would always start searching for the package from the root `node_modules`. That problem is fixed but there's still cases where people don't delete old `node_modules` directories and sdk resolves the wrong sdk/core version. Since the upgrade we had multiple cases of these errors and it was hard to debug, that's how this PR came to life. 

About the versioning: is there a package that we already use or should I install one. I knew my `includes` implementation is too simple but I just wanted to get this conversation started.

Reply:
@henrist I completely understand your point. But while upgrading the `@cognite/react-container` package to use the new SDK version I encountered this problem. The reason was multiple `tsconfig` settings and resolutions in `package.json`. It took me so long to find out what was going on because my editor would point me to the correct version, but when the code would run it would resolve an old one. And I've seen this behaviour with apps outside of the monorepo too because they all could have different configurations and if they don't clean the `node_modules` but just install it doesn't resolve the core version correctly. 

That is the reason why I think something like this that checks at runtime could be a helpful thing. But if you think that we don't want this then I'll close the PR

Reply:
I don't think we should close this. This work does fix _a_ problem, though potentially not all _problems_ as you pointed out. Either way it would still be beneficial for a certain set of users/situations in close to the current state.

Reply:
Updated the PR to use `semver` for version check

Reply:
@henrist Do you think it is ok now, can we merge this in ?